runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -1452,15 +1452,25 @@ TR::Register *J9::X86::TreeEvaluator::newEvaluator(TR::Node *node, TR::CodeGener
    TR::Compilation *comp = cg->comp();
    TR::Register *targetRegister=NULL;
 
-   // If the helper symbol set on the node is TR_newValue, we are (expecting to be)
-   // dealing with a value type. Since we do not fully support value types yet, always
-   // call the JIT helper to do the allocation.
-   //
-   if (TR::Compiler->om.areValueTypesEnabled() && node->getSymbolReference() == comp->getSymRefTab()->findOrCreateNewValueSymbolRef(comp->getMethodSymbol()))
+   if (TR::Compiler->om.areValueTypesEnabled() && node->getOpCodeValue() == TR::New)
       {
-      TR_OpaqueClassBlock *classInfo;
-      bool spillFPRegs = comp->canAllocateInlineOnStack(node, classInfo) <= 0;
-      return TR::TreeEvaluator::performHelperCall(node, NULL, TR::acall, spillFPRegs, cg);
+      TR::SymbolReference *newValueSymRef = comp->getSymRefTab()->findOrCreateNewValueSymbolRef(comp->getMethodSymbol());
+      TR::SymbolReference *nodeSymRef = node->getSymbolReference();
+      TR::SymbolReference *classSymRef = node->getFirstChild()->getSymbolReference();
+
+      TR_OpaqueClassBlock *clazz = NULL;
+      if (!classSymRef->isUnresolved())
+         clazz = (TR_OpaqueClassBlock *)classSymRef->getSymbol()->getStaticSymbol()->getStaticAddress();
+
+      bool isValueTypeClass = clazz ? TR::Compiler->cls.isValueTypeClass(clazz) : false;
+
+      if (((nodeSymRef == newValueSymRef) && !isValueTypeClass)
```

Comment:
I think a comment here would be helpful.  It might not be obvious that this is checking for the cases where `new jitNewValue` is being used for an identity class or `new jitNewObject` is being used for a value-type class, and forcing it to use the helper call so an exception can be thrown.

Reply:
Addressed in b796c85.